strawberryfields/program.py

```diff
@@ -529,17 +544,7 @@ def _get_compiler(compiler_or_name):
                 compiler = _get_compiler(compiler)
 
             if device.modes is not None:
-                # Check that the number of modes in the program is valid for the given device.
-
-                # Program subsystems may be created and destroyed during execution. The length
-                # of the program registers represents the total number of modes that has ever existed.
-                modes_total = len(self.reg_refs)
-
-                if modes_total > device.modes:
-                    raise CircuitError(
-                        f"This program contains {modes_total} modes, but the device '{target}' "
-                        f"only supports a {device.modes}-mode program."
-                    )
+                self.assert_number_of_modes(device)
```

Comment:
@josh146 @antalszava Just realized that the recently merged [No modes check for simulon PR](https://github.com/XanaduAI/strawberryfields/pull/482) clashes with this one, since that PR added a check for `isinstance(device.modes, int)` rather than simply checking for `device.modes != "simulon"` (as was first proposed). That doesn't work with the changes here, so we'd have to update it by e.g. moving it to inside the `assert_number_of_modes` method, or changing the check (since not only integers should eval `True` here.

Reply:
Hmm, if I understand correctly, the clash is related to a `TDMProgram` inheriting this method, correct? Would it be solved by having `TDMProgram.compile` be redefined in #484? :thinking: We also have #483 open

Reply:
Oh, that's true. We should add a call to `assert_number_of_modes` in `TDMProgram.compile` as well then, and perhaps just leave the `isinstance(device.modes, int)` here. I feel like this is becoming a bit confusing though, but we're planning on adding checks for other mode entries (e.g. dicts) as well, right? In that case perhaps it would be best just to expand the `assert_number_of_modes` method to check for all possibilities.

Reply:
Yes, I think you're right. The way I understand how a potential way could be:
1. Wait for #484 to be merged in here *or* create a separate condition to resolve conflict in `program.py` and allow the tests to pass e.g.,
```python
            if device.modes is not None and (isinstance(device.modes, int) or device.target == "tdm"):
                self.assert_number_of_modes(device)
```
2. Once #484 was merged, if the new condition in 1. was added, remove `or device.target == "tdm"`
3. When resolving #483, expand `assert_number_of_modes` to check for `dict` types